Adding functionality to initialise RunReaders at offset

recursion-ninja opened the pull request:
# Description

First part of Work Product 12, adding the functionality to start `RunReaders` at an offset.

The new `rawPageFindKey` function has the same performance characteristics as `rawPageLookup` but facilitates starting the `RunReaders` at an give page offset from the returned index. `RunReader.new` now takes an optional offset as an argument.

jorisdral commented on test/Test/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -294,7 +296,7 @@ prop_big_insert k mblobref =
     v       = Ref.Value (BS.replicate 5000 42)
     op      = Ref.Insert v mblobref
-    kops    = [(k, op)]
+    kops    = [ (k, op)]
```
```suggestion
    kops    = [(k, op)]
```

jorisdral commented on test/Test/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -187,6 +187,8 @@ tests = testGroup "Database.LSMTree.Internal.RawPage"
     , testProperty "entry" prop_single_entry
     , testProperty "rawPageOverflowPages" prop_rawPageOverflowPages
     , testProperty "from/to reference impl" prop_fromToReferenceImpl
+    , testProperty "∀ key page. maybe True (key <=) (getRawPageIndexKey . rawPageIndex page .  id  =<< rawPageFindKey page key)" prop_findKey_index_law_GT
+    , testProperty "∀ key page. maybe True (key > ) (getRawPageIndexKey . rawPageIndex page . pred =<< rawPageFindKey page key)" prop_findKey_index_law_LT
```
One argument in favour of simpler test names, e.g. just `prop_findKey_index_law_LT`, is that it's easier to search the code for these names, and it's easier to run the tests in isolation using `-p`. What are your thoughts?

I often just put the name of the function in the tests, and if I want to describe what a test does more succinctly then I'll put it as a haddock comment on the function

recursion-ninja replied:
In my own projects, for type-class laws, I have included the very verbose name of the law. However, I think that since this project has consistently used shorter test names I should modify these test names to follow suit. 

I think I'd like to, if you don't object, move the verbose "relational law" into the counter example message which appears only if the property fails to hold.

jorisdral commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -187,10 +209,28 @@ next fs hbio reader@RunReader {..} = do
 readDiskPage :: HasFS IO h -> FS.Handle h -> IO (Maybe RawPage)
 readDiskPage fs h = do
     mba <- newPinnedByteArray pageSize
+    -- TODO: make sure no other exception type can be thrown
```
I don't think you can prevent this

recursion-ninja replied:
Yeah, I put that note in more as a reminder to review this area later.

jorisdral commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -64,18 +65,39 @@ data RunReader s fhandle = RunReader {
+        firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs readerKOpsHandle
+        return (firstPage, 0)
+
+      Just offset -> do
+        let pageNo = Index.pageSpanStart (Index.search offset index)
+        -- load first page from disk, if it exists.
+        firstPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+
+        case rawPageFindKey firstPage offset of
+          Just n ->
+            return (firstPage, n)
+          Nothing -> do
+            -- actually, it's the next page
+            nextPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+            return (nextPage, 0)
```
We use `readDiskpage` in `next`, which means that we'll read the first page at an offset, but then start reading pages from the start. This is probably not caught by tests because we never start reading at an offset. We should extend the tests we already have to ensure we're testing this case properly

mheinzel replied:
Ah, yes, the handle still points at the beginning, because `hGetBufExactlyAt` does not change the handle state. Maybe we should use `hSeek` instead?

Alternatively, we could keep track of the page number ourselves and always use `hGetBufExactlyAt`.

mheinzel replied:
> you're going to have to either systematically use `readDiskPageAt` or use seeking

I don't think this has been resolved yet @recursion-ninja.

> Using `readDiskPageAt` and keeping track ourselves would get us closer to being able to use the same `O_DIRECT` file handle as the lookups use (we'd have to add our own manual read-ahead)

Do we want this? We save a file handle, but creating a new one for sequential reading and simply relying on the OS to do read-ahead seems quite convenient.

mheinzel replied:
This has been resolved by seeking, so we still rely on the handle's internal file pointer.

jorisdral commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -186,6 +246,18 @@ data RawPageIndex entry =
+-- |
+-- __Time:__ \( \mathcal{O}\left( 1 \right) \)
+--
+-- Conveniently access the the 'SerialisedKey' of a 'RawPageIndex'.
```
```suggestion
-- Conveniently access the 'SerialisedKey' of a 'RawPageIndex'.
```

jorisdral reviewed (commented):
This is not a full review yet -- here are some initial comments

jorisdral reviewed (changes requested):
Whoops, forgot to request changes

mheinzel commented:
There is a [build failure for GHC versions before 9.6](https://github.com/IntersectMBO/lsm-tree/actions/runs/10493900676/job/29068935700?pr=346#step:13:151) which seems related to `UnboxedSums`.

Also, there is a [docspec-related error](https://github.com/IntersectMBO/lsm-tree/actions/runs/10493900676/job/29068937398?pr=346#step:17:103), seems to happen on all Ubuntu builds.

mheinzel commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
+-- where \( n \) is the number of entries in the 'RawPage'.
+--
+-- Return the least entry number in the 'RawPage' (if it exists)
+-- which is greater than or equal to the suppled 'SerialisedKey'.
```
```suggestion
-- which is greater than or equal to the supplied 'SerialisedKey'.
```

mheinzel commented on test/Test/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -187,6 +187,8 @@ tests = testGroup "Database.LSMTree.Internal.RawPage"
     , testProperty "entry" prop_single_entry
     , testProperty "rawPageOverflowPages" prop_rawPageOverflowPages
     , testProperty "from/to reference impl" prop_fromToReferenceImpl
+    , testProperty "∀ key page. maybe True (key <=) (getRawPageIndexKey . rawPageIndex page .  id  =<< rawPageFindKey page key)" prop_findKey_index_law_GT
+    , testProperty "∀ key page. maybe True (key > ) (getRawPageIndexKey . rawPageIndex page . pred =<< rawPageFindKey page key)" prop_findKey_index_law_LT
```
These tests only assert anything when `rawPageFindKey` returns `Just`. Can you also add a property that whenever the last key of the page is `(>= key)`, then we successfully find an index number?

Looking at the code I have a hunch that it might sometimes incorrectly return `Nothing`, but I could be wrong.

recursion-ninja replied:
@mheinzel Your hunch seems to have been accurate. I added a property test for the `Nothing` case and it does fail! Any thoughts on when it incorrectly returns `Nothing`?

recursion-ninja replied:
You can replicate a failure case on this branch by invoking:

```
cabal test lsm-tree-test --test-options='--quickcheck-replay="(SMGen 3937336630516587970 3943216271980923571,0)" -p "/rawPageFindKey missing/"'
```

recursion-ninja replied:
@mheinzel I have figured out the corner case in the binary search!

mheinzel replied:
Nice!

The properties we seem a little complex by now. On the other hand, it is quite straightforward to to create a reference implementation working on the list of k/ops, similar to the test for `rawPageLookup`. Maybe that would be a nicer approach?

dcoutts commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -187,10 +209,28 @@ next fs hbio reader@RunReader {..} = do
+readDiskPageAt :: HasFS IO h -> Index.PageNo -> FS.Handle h -> IO (Maybe RawPage)
+readDiskPageAt fs pageNo h = do
+    mba <- newPinnedByteArray pageSize
+    -- TODO: make sure no other exception type can be thrown
```
Others will just propagate. Is that ok?

recursion-ninja replied:
After having some weeks to think about it, I don't think we can properly handle other exceptions at this call site.

recursion-ninja replied:
@dcoutts I have added a paraphrased version of your commentary as a comment to `RunReader.new`.

dcoutts commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
+-- NOTE: Be sure to enable UnboxedSums to improve efficiency of this data-type.
+data BinarySearchResult
+  = KeyNotInPage
+  | KeyFoundAt {-# UNPACK #-} !Ordering  {-# UNPACK #-} !Int
```
The warnings here are correct, you cannot unpack the `Ordering`, but on the other hand given that it's an enum, there's little advantage either.

Note also that we only pattern match on EQ or not EQ, so we don't need the full `Ordering`. A bool would be enough.

dcoutts commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+--
+-- NOTE: Be sure to enable UnboxedSums to improve efficiency of this data-type.
```
Is this a TODO? Enabling `UnboxedSums` will not do anything automatically, you'd have to explicitly make `bisectPageToKey` return an unboxed sum. There's no worker/wrapper equivalent to do this automagically, AFAIK.

dcoutts commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
-        k = i + div2 (j - i)
-
-    linear :: Int -> Int -> RawPageLookup (Entry SerialisedValue BlobSpan)
+    -- k in [i, j)
+    -- k >= key
+    -- k-1 < key
```
Presumably these are preconditions? Say so explicitly.

dcoutts commented on test/Test/Database/LSMTree/Internal/Run.hs:
```diff
@@ -262,7 +262,7 @@ isLargeKOp (key, entry) = size > pageSize
 readKOps :: FS.HasFS IO h -> FS.HasBlockIO IO h -> Run RealWorld (FS.Handle h) -> IO [SerialisedKOp]
 readKOps fs hbio run = do
-    reader <- Reader.new fs hbio run
+    reader <- Reader.new fs hbio Nothing run
```
Will we have tests for the new functionality later? :-) In this PR or another PR?

recursion-ninja replied:
Another PR I think. I'll need some guidance on the semantics of how to set up an "imperative" test correctly with our helper functions.

mheinzel replied:
There are tests for this now: https://github.com/IntersectMBO/lsm-tree/pull/346/files#diff-48a02884e73d81f478092b7cb891f5af973b8b6844257c98f597415d0652e159

dcoutts reviewed (changes requested):
This is looking good. No major grumbles.

recursion-ninja commented on src/Database/LSMTree/Internal/RunReaders.hs:
```diff
@@ -91,7 +92,26 @@ new fs hbio runs = do
+-- | On equal keys, elements from runs earlier in the list are yielded first.
+-- This means that the list of runs should be sorted from new to old.
+-- TODO: merge with 'new'?
+newAtOffset :: forall h .
```
I transferred the functionality to a generalized `newAtOffsetMaybe` function. Both `new` and `newAtOffset` call `newAtOffsetMaybe` f while being able to retain their original type signatures (no `Maybe` in `newAtOffset`).

dcoutts commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -71,26 +71,41 @@ new ::
+
+          _ -> do
+            -- Otherwise, we return the beginning of the next page.
+            nextPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs (Index.nextPageNo pageNo) readerKOpsHandle
```
So this tries to handle the case where there is no next page, because we were already at the last page (corresponding to asking for a greater than all other keys).

I see that `readDiskPageAt` looks for the `FsReachedEOF` exception. Do we know it definitely works that way when using the "at" functions, like `hGetBufExactlyAt` to request something past the end of the file?

I think it's worth adding a unit test (or a label in a property test) to make sure we cover the case of asking for a key that's past the end of all the keys in the db, so we cover this case here.

mheinzel replied:
I would also like to see some tests to be confident about the correctness. It's a bit tricky.

For example, I think `nextPageNo pageNo` here doesn't work if there is a multi-page value. We'd end up loading an overflow page. It should be `nextPageNo pageEnd` instead.

mheinzel replied:
One way to get some coverage would be to add `newAtOffset` to the tests in `Test.Database.LSMTree.Internal.RunReaders`.

dcoutts replied:
> For example, I think nextPageNo pageNo here doesn't work if there is a multi-page value.

Ah yes, you're right.

recursion-ninja replied:
Ah, I understand better now. I changed the parameter to `Index.nextPageNo pageEnd`.

Just to be certain, it is *unconditionally* `Index.nextPageNo pageEnd`, or do we need to check for a multi-page value and branch between `Index.nextPageNo pageEnd` and `Index.nextPageNo pageNo`?

mheinzel replied:
Unconditionally, since for single-page values `pageNo == pageEnd`.

dcoutts commented on src/Database/LSMTree/Internal/RunReaders.hs:
```diff
@@ -86,15 +86,7 @@ new :: forall h .
-  where
-    fromRun :: ReaderNumber -> Run IO (FS.Handle h) -> IO (Maybe (ReadCtx (FS.Handle h)))
-    fromRun n run = nextReadCtx fs hbio n =<< Reader.new fs hbio Nothing run
+new = newAtOffsetMaybe Nothing
```
:+1:

dcoutts reviewed (approved):
LGTM.

Once you're happy enough reviews are in, squash these commits together.

mheinzel commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
+
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+data BinarySearchResult
+  = KeyNotInPage
+  | KeyFoundAt !Bool {-# UNPACK #-} !Int
```
The Bool here is not explained and generally I think the constructor names are a bit misleading. `KeyNotInPage` means specifically that the key is larger than the largest key in the page. `KeyFoundAt False i` means that the key was *not* found, but the next-largest key can be found at index `i`.

Maybe three constructors would be clearer? (I think the allocation should still be optimised away, but might be worth checking)

```haskell
  = KeyIsTooLarge
  | KeyNotFoundNextIs {-# UNPACK #-} !Int
  | KeyFoundAt {-# UNPACK #-} !Int
```

We can also leave it at two constructors, but then tweaking the name and commenting generously would be nice.

recursion-ninja replied:
I expanded the sum type to be more explicit.

```haskell
data BinarySearchResult                                                                                                         
  = KeyNotFoundExceedsPage                                                                                                      
  | KeyNotFoundNearestIsAt {-# UNPACK #-} !Int                                                                                  
  | KeyWasFoundAt {-# UNPACK #-} !Int 
```

mheinzel replied:
`Next` would be a bit clearer than `Nearest`. The previous key might be very "near", but we still return the next one.

mheinzel commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
-        | i >= j                     = LookupEntryNotPresent
-        | key == rawPageKeyAt page i = LookupEntry (rawPageEntryAt page i)
-        | otherwise                  = linear (i + 1) j
+      | i > j || (i == j && i == quantity) = KeyNotInPage
```
`i > j` should never happen, so no need to check it. We could add `i <= j` as an invariant in the comment above and also an assertion.

Also, if `i == j` but `i /= quantity`, we know immediately that `KeyFoundAt False j` should be returned (the range `[i, j)` is empty), no need to call `rawPageKeyAt` again.

So overall this might look like `| i == j = if i == quantity then KeyNotInPage else KeyFoundAt False j`.

recursion-ninja replied:
You are correct, that is equivalent and simpler logic. I will leave the `i >= j` for the sake of ensuring totality of the stopping condition though. You never know when a cosmic ray might flip a bit and we need the loop to stop 😉.

mheinzel commented on src/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
+  -> BinarySearchResult
+bisectPageToKey !page !key = go 0 quantity
+  where
+    quantity = fromIntegral $ rawPageNumKeys page
```
Something like `numKeys` would make it clearer which quantity is being referred to.

recursion-ninja replied:
Done!

mheinzel commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -64,18 +65,54 @@ data RunReader m fhandle = RunReader {
+          -- The information tells us that the key does not exist, but that if it were to exist,
+          -- it would be between the last key in this page and the first key in the next page.
+          -- Thus the reader should be initialised to return keys starting from the next page.
+          Nothing | pageNo == pageEnd ->
+            -- If already at the last page, there is no page of content to return.
+            -- Hence, no need to call readDiskPageAt.
+            return (emptyRawPage, 0)
```
I don't understand this case. Do you mean something like `pageNo >= sizeInPages index`? 

`pageEnd` is not the last page of the entire run, just the last page of the (potential) entry we are looking for. So `pageNo == pageEnd` means that we don't have to deal with a larger-than-page entry here.

mheinzel reviewed (commented):
Still a few things that should be addressed IMO, mainly the issues around keys that fall within pages (where `rawPageFindKey` returns `Nothing`) and the usage of `readDiskPage` which currently doesn't take the offset into account properly.

mheinzel commented on test/Test/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
+              | keySmall == keyLarge = []
+              | otherwise =
+                let missingKey = do
+                      let n = length keys - 1
+                      afterIndex <- (`mod` n) <$> arbitrary
```
Seems like a job for [chooseInt](https://www.stackage.org/haddock/lts-22.8/QuickCheck-2.14.3/Test-QuickCheck.html#v:chooseInt).

mheinzel commented on test/Test/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
+--
+-- The first key is assumed to be the "smaller" of the pair.
+--
+-- NOTE: Under pathological input conditions, there is an infintesimally small chance
```
```suggestion
-- NOTE: Under pathological input conditions, there is an infinitesimally small chance
```

mheinzel commented on test/Test/Database/LSMTree/Internal/RawPage.hs:
```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
+-- NOTE: Under pathological input conditions, there is an infintesimally small chance
+-- that this will generate /the same/ key as the larger value. This will not effect the
+-- validity of 'Arbitrary' instances which depend on this function.
+arbitrarySerialisedKeyBetween :: SerialisedKey -> SerialisedKey -> Gen SerialisedKey
+arbitrarySerialisedKeyBetween (SerialisedKey (RawBytes loBytes)) (SerialisedKey (RawBytes hiBytes)) =
+    SerialisedKey . RawBytes <$> VP.zipWithM withinByte loBytes hiBytes
+  where
+    withinByte !lhs !rhs
+      | lhs == rhs = pure rhs
+      | otherwise =
+        let (!lo, !hi)
+              | lhs < rhs = (lhs, rhs)
+              | otherwise = (rhs, lhs)
+            !diff = hi - lo
+        in  (succ lo +) . (`mod` diff) <$> arbitrary
```
This does a slightly weird thing, where *each byte* is between the corresponding bytes of the two keys. This limits the set of keys that get generated quite a bit. If the first byte is already smaller/bigger than the first byte of the lower/upper bound, then the remaining bytes could be completely arbitrary.

mheinzel replied:
We can leave this as is, since the only thing observed about keys is their ordering towards each other. Still feels a bit off.

A potential solution would be to use a simpler generator (maybe only 50:50 existing key or arbitrary key), with labels for the cases specifically targetted here, to make sure the distribution looks good. Recognizing a property of the input is usually simpler than having to specifically generate input with that property.

recursion-ninja replied:
I took your advice to just simplify the generator and change the distribution to choose an existing key with $\mathbf{P}\left\[\frac{1}{2}\right\]$ and an `arbitrary` key otherwise.

recursion-ninja replied:
Actually, I've reversed my position in light of new information. I annotated the tests with [`label`](https://hackage.haskell.org/package/tasty-quickcheck/docs/Test-Tasty-QuickCheck.html#v:label) to see how the 50/50 distribution of random/existing key selection panned out. There was *essentially* no coverage of the "before page" and "after page" cases. So I decided to bring back the computationally more complicated, but more predictable distribution of outcomes.

After bringing back the original, more complex distribution, I updated the sub-generators to have an expanded value range. The updated ranges address the previously "weird" location-independent byte generation which you had commented on. 

I had chosen that admittedly weird generation technique to simplify the generation process. To correctly generate any key within an upper and/or lower byte-stream, I would need to track if I had generated a smaller/larger byte earlier in the sequence. This required dragging state through the QuickCheck generator which I had lazily avoided doing. That state dragging has been added now.

mheinzel replied:
The usage of `min`/`max` still doesn't seem 100% right, but it should be good enough and lead to a good distribution of test cases.

mheinzel commented on test/Test/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -0,0 +1,216 @@
+                     withTempIOHasBlockIO)
+import           Test.Util.Orphans ()
+
+-- TODO: this is pretty redundant with the tests for 'Run.Run'. Deduplicate?
```
Has been de-duplicated, so the TODO can be dropped.

mheinzel reviewed (commented):
(seems like I accidentally submitted parts of the review already)

The `RawPage` tests still seem a bit more complicated than they could be (see comments about generator, also [comparing against a reference implementation](https://github.com/IntersectMBO/lsm-tree/pull/346#discussion_r1742246587) could be simpler than the three separate properties?). This doesn't necessarily have to be a blocker, though.

However, it seems like `rawPageLookup` now allocates more, which should be investigated.

recursion-ninja commented:
## Regression Fix Report

I have isolated and corrected the allocation regression in `rawPageLookup` that the PR contained. The excess allocation came from the "linear scanning" portion of the `bisectPageToKey` function. 

Previously, when the bench-marking showed identical allocations and comparable run-times, there was a defect in the `bisectPageToKey` function. To correct the defect, an additional check was required to determine of the binary search ended at the last entry in the page. If so, the binary search result needed to be changed from "found closest key at index `i`" to "key not found in page." This additional comparison work, which existed in the terminal "linear scan" portion of the binary search routine, was inexorably linked to the additional allocation. 

I attempted to alter the code while retaining the linear scan portion of the binary search routine. After repeated an persistent failure to remove the additional allocation, I removed the linear scan segment of the binary search routine.  To my pleasant surprise, the allocations decreased to the same *or less than* the allocations reported on `main`!

Below is a benchmark table I collated describing the change in bytes allocated (B) and run-time measured (ns) between `main` and this PR across recent versions of GHC. In the Δ column, a negative number represents an *improvement*  when measuring the performance of this PR (relative to `main`).

## Results

| `rawPageLookup`  | GHC        | `main` (B, ns)    | PR-#346 (B, ns)   | Diif-Δ (B, ns)    |
|:-----------------|:-----------|------------------:|------------------:|------------------:|
| `missing`        | **9.10.1** | (`544`, `101.50`) | (`544`, `104.90`) | (`± 0`, `+ 3.40`) |
| `existing-head`  | **9.10.1** | (`560`, ` 98.72`) | (`560`, `101.70`) | (`± 0`, `+ 2.98`) |
| `existing-last`  | **9.10.1** | (`640`, ` 96.91`) | (`576`, ` 88.27`) | (`-64`, `- 8.64`) |
| `missing`        | **9.8.2**  | (`544`, `102.90`) | (`544`, `104.90`) | (`± 0`, `+ 2.00`) |
| `existing-head`  | **9.8.2**  | (`560`, `100.40`) | (`560`, `103.00`) | (`± 0`, `+ 2.60`) |
| `existing-last`  | **9.8.2**  | (`640`, ` 98.92`) | (`576`, ` 91.17`) | (`-64`, `- 7.75`) |
| `missing`        | **9.6.6**  | (`544`, `104.20`) | (`544`, `103.50`) | (`± 0`, `- 0.70`) |
| `existing-head`  | **9.6.6**  | (`560`, ` 99.82`) | (`560`, ` 90.30`) | (`± 0`, `- 9.52`) |
| `existing-last`  | **9.6.6**  | (`640`, ` 97.21`) | (`576`, ` 88.48`) | (`-64`, `- 9.73`) |
| `missing`        | **9.4.8**  | (`544`, `102.90`) | (`544`, `105.40`) | (`± 0`, `+ 2.50`) |
| `existing-head`  | **9.4.8**  | (`560`, `100.00`) | (`560`, `103.10`) | (`± 0`, `+ 3.10`) |
| `existing-last`  | **9.4.8**  | (`640`, ` 99.06`) | (`576`, ` 88.84`) | (`-64`, `-10.22`) |

## Conclusion

My reading of the tabulated measurements is that the PR, with the recent regression fix, now constitutes an *unequivocal improvement* in **allocations** and a *minor to negligible improvement* in **run-time**!

recursion-ninja commented:
I believe that this resolves the last critical technical issue of the PR. I'm going to give the PR a final review to ensure that all comments have been addressed.

mheinzel reviewed (approved):
Good working on tracking down the allocation regression. I think everything has been addressed now.
